aptos-move/aptos-vm-types/src/module_and_script_storage/state_view_adapter.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::module_and_script_storage::{
+    module_storage::AptosModuleStorage, script_storage::AptosScriptStorage,
+};
+use aptos_types::state_store::{state_key::StateKey, StateView};
+use bytes::Bytes;
+use move_binary_format::{errors::PartialVMResult, file_format::CompiledScript, CompiledModule};
+use move_core_types::{account_address::AccountAddress, identifier::IdentStr, metadata::Metadata};
+use move_vm_runtime::{
+    module_storage_error, move_vm::MoveVM, IntoUnsyncCodeStorage, Module, ModuleBytesStorage,
+    ModuleStorage, Script, ScriptStorage, UnsyncCodeStorage, UnsyncModuleStorage,
+};
+use std::sync::Arc;
+
+/// Avoids orphan rule to implement [ModuleBytesStorage] for [StateView].
+struct StateViewAdapter<'s, S> {
+    state_view: &'s S,
+}
+
+impl<'s, S: StateView> ModuleBytesStorage for StateViewAdapter<'s, S> {
+    fn fetch_module_bytes(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> PartialVMResult<Option<Bytes>> {
+        let state_key = StateKey::module(address, module_name);
+        self.state_view
+            .get_state_value_bytes(&state_key)
+            .map_err(|e| module_storage_error!(address, module_name, e))
+    }
+}
+
+/// A (not thread-safe) implementation of code storage on top of a state view.
+/// It is never built directly by clients - only via [AsAptosCodeStorage] trait.
+/// Can be used to resolve both modules and scripts.
+pub struct AptosCodeStorageAdapter<'s, S> {
+    storage: UnsyncCodeStorage<UnsyncModuleStorage<'s, StateViewAdapter<'s, S>>>,
+}
+
+impl<'s, S: StateView> AptosCodeStorageAdapter<'s, S> {
+    fn new(state_view: &'s S, vm: &'s MoveVM) -> Self {
+        let adapter = StateViewAdapter { state_view };
+        let storage = adapter.into_unsync_code_storage(vm.runtime_env());
+        Self { storage }
+    }
+}
+
+impl<'s, S: StateView> ModuleStorage for AptosCodeStorageAdapter<'s, S> {
+    fn check_module_exists(
```

Comment:
can we use this (https://crates.io/crates/delegate) or some similar crate (or macros if not) to reduce the boilerplate?

Reply:
That one does not work with traits, but I ll try to use this one: https://crates.io/crates/ambassador

Reply:
Fixed in #14402